src/stochastic_swap.rs

```diff
@@ -0,0 +1,352 @@
+
+    while depth_step < depth_max {
+        let mut qubit_set = input_qubit_set.clone();
+        while !qubit_set.is_empty() {
+            min_cost = compute_cost(&scale.view(), &trial_layout, gates, num_gates);
+            // Try to decrease the objective function
+            cost_reduced = false;
+            for idx in 0..num_edges {
+                let start_edge = edges[2 * idx];
+                let end_edge = edges[2 * idx + 1];
+                let start_qubit = trial_layout.phys_to_logic[start_edge];
+                let end_qubit = trial_layout.phys_to_logic[end_edge];
+                if qubit_set.contains(&start_qubit) && qubit_set.contains(&end_qubit) {
+                    // Try this edge to reduce cost
+                    trial_layout.swap(start_edge, end_edge);
+                    // compute objective function
+                    new_cost = compute_cost(&scale.view(), &trial_layout, gates, num_gates);
+                    // record progress if we succeed
+                    if new_cost < min_cost {
+                        cost_reduced = true;
+                        min_cost = new_cost;
+                        optimal_layout = trial_layout.clone();
+                        optimal_start = start_edge;
+                        optimal_end = end_edge;
+                        optimal_start_qubit = start_qubit;
+                        optimal_end_qubit = end_qubit;
+                    }
+                    trial_layout.swap(start_edge, end_edge);
+                }
+            }
+            // After going over all edges
+            // Were there any good swap choices?
+            if cost_reduced {
+                qubit_set.remove(&optimal_start_qubit);
+                qubit_set.remove(&optimal_end_qubit);
+                trial_layout = optimal_layout.clone();
+                opt_edges.add(optimal_start, optimal_end);
+            } else {
+                break;
+            }
+        }
+        // We have either run out of swap pairs to try or failed to improve
+        // the cost
+
+        // Compute the coupling graph distance
+        dist = compute_cost(&cdist, &trial_layout, gates, num_gates);
+        // If all gates can be applied now we're finished.
+        // Otherwise we need to consider a deeper swap circuit
+        if dist as usize == num_gates {
+            break;
+        }
+        // increment the depth
+        depth_step += 1;
+    }
+    // Either we have succeeded at some depth d < d_max or failed
+    dist = compute_cost(&cdist, &trial_layout, gates, num_gates);
+    if let Some(locked_best_possible) = locked_best_possible {
+        if dist as usize == num_gates && depth_step == 1 {
+            let mut best_possible = locked_best_possible.write().unwrap();
+            // In the case a ideal solution has already been found to pr
+            // behavior consistent with the single threaded predecessor to this function
+            // we defer to the earlier trial
```

Comment:
looks like the first line of this comment got truncated

Reply:
Fixed in: https://github.com/Qiskit/qiskit-terra/pull/7658/commits/eaf92f2199b4a14472ad2b8a6155a29c1cc5b782